File: AdaptiveTD/AdaptiveTD/AdaptiveTD/Missile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace AdaptiveTD
{
    class Missile
    {
        Texture2D selfTexture;
        Vector2 velocity, position, oldPosition;
        float travelledDistance, distanceToTarget;
        public bool remove;
        int damage;
        Enemy targetEnemy;
        DamageOverTime damageOverTime;
        Slow slow;
        AreaOfEffect areaOfEffect;

        public Missile(Texture2D selfTexture, Vector2 start, float distanceToTarget, float angle, float totalVelocity, Enemy targetEnemy, int damage, DamageOverTime damageOverTime, Slow slow, AreaOfEffect areaOfEffect)
        {
            this.selfTexture = selfTexture;
            position = start;
            this.distanceToTarget = distanceToTarget;
            this.travelledDistance = 0.0f;
            this.remove = false;
            this.targetEnemy = targetEnemy;
            this.damage = damage;
            this.damageOverTime = damageOverTime;
            this.slow = slow;
            this.areaOfEffect = areaOfEffect;

            velocity.X = (float)(totalVelocity * Math.Cos(angle));
            velocity.Y = (float)(totalVelocity * Math.Sin(angle));
        }

        public void Update(float gameTime, List<Enemy> enemies)
        {
            oldPosition = position;
            position += velocity * (float)gameTime;
            travelledDistance += (float)(Math.Sqrt(Math.Pow(position.X - oldPosition.X, 2) + Math.Pow(position.Y - oldPosition.Y, 2)));
            if (travelledDistance >= distanceToTarget)
            {
                remove = true;
                targetEnemy.Health -= damage;
                targetEnemy.DamageOverTime = damageOverTime;
                targetEnemy.Slow = slow;
                if (areaOfEffect.Radius > 0)
                    DoAoE(areaOfEffect.Radius, damage, enemies);
            }
        }

        private void DoAoE(float radius, int damage, List<Enemy> enemies)
        {
            foreach (Enemy candidateEnemy in enemies) //Algoritme kan improves ved å ikke teste på første element
            {
                double deltaXCandidate = (candidateEnemy.Position.X + candidateEnemy.Origin.X) - (this.position.X);
                double deltaYCandidate = (candidateEnemy.Position.Y + candidateEnemy.Origin.Y) - (this.position.Y);
                double distanceToCandidate = Math.Sqrt(Math.Pow(deltaXCandidate, 2) + Math.Pow(deltaYCandidate, 2));

                if (distanceToCandidate <= radius)
                {
                    candidateEnemy.Health -= damage;
                }
            }
        }

        public void Draw(SpriteBatch spriteBatch)
        {
            spriteBatch.Draw(selfTexture, position, Color.White);
        }
    }
}

File: LibGDXProjects/AdaptiveTD/src/no/uia/adaptiveTD/Missile.java
package no.uia.adaptiveTD;

import java.util.HashMap;
import java.util.List;

import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.Vector2;

public class Missile {
	HashMap<Integer, Sprite> sprites;
    Vector2 velocity, position, oldPosition;
    float travelledDistance, distanceToTarget;
    public Boolean remove;
    int damage;
    Enemy targetEnemy;
    
    public Missile(Sprite missileSprite, Vector2 start, float distanceToTarget, float angle, float totalVelocity, Enemy targetEnemy, int damage){
        position = start;
        this.distanceToTarget = distanceToTarget;
        this.travelledDistance = 0.0f;
        this.remove = false;
        this.targetEnemy = targetEnemy;
        this.damage = damage;
        this.sprites = new HashMap<Integer, Sprite>();
        this.sprites.put(0, missileSprite);
        velocity.x = (float)(totalVelocity * Math.cos(angle));
        velocity.y = (float)(totalVelocity * Math.sin(angle));
    }

    public void Update(float gameTime, List<Enemy> enemies)
    {
        oldPosition = position;
        position.x += velocity.x * (float)gameTime;
        position.y += velocity.y * (float)gameTime;
        travelledDistance += (float)(Math.sqrt(Math.pow(position.x - oldPosition.x, 2) + Math.pow(position.y - oldPosition.y, 2)));
        if (travelledDistance >= distanceToTarget)
        {
            remove = true;
            targetEnemy.currentHealth -= damage;
            /*
            targetEnemy.DamageOverTime = damageOverTime;
            targetEnemy.Slow = slow;
            if (areaOfEffect.Radius > 0)
                DoAoE(areaOfEffect.Radius, damage, enemies);
                */
        }
    }

    private void DoAoE(float radius, int damage, List<Enemy> enemies)
    {
        for (Enemy candidateEnemy : enemies) //Algoritme kan improves ved å ikke teste på første element
        {
            double deltaXCandidate = (candidateEnemy.getPosition().x + candidateEnemy.getOrigin().x) - (this.position.x);
            double deltaYCandidate = (candidateEnemy.getPosition().y + candidateEnemy.getOrigin().y) - (this.position.y);
            double distanceToCandidate = Math.sqrt(Math.pow(deltaXCandidate, 2) + Math.pow(deltaYCandidate, 2));

            if (distanceToCandidate <= radius)
            {
                candidateEnemy.currentHealth -= damage;
            }
        }
    }

    public void Draw(SpriteBatch spriteBatch)
    {
        spriteBatch.draw(sprites.get(0), position.x, position.y);
    }

}

 

